homeassistant/components/habitica/config_flow.py

```diff
@@ -0,0 +1,107 @@
+"""Config flow for habitica integration."""
+import logging
+
+import requests
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import CONF_API_KEY, CONF_NAME, CONF_URL, HTTP_OK
+
+from .const import (  # pylint: disable=unused-import
+    API_STATUS,
+    API_USER_PROFILE,
+    CONF_API_USER,
+    DEFAULT_URL,
+    DOMAIN,
+)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_USER): str,
+        vol.Required(CONF_API_KEY): str,
+        vol.Optional(CONF_NAME): str,
+        vol.Optional(CONF_URL, default=DEFAULT_URL): str,
+    }
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect."""
+
+    def check_api_status(url=DEFAULT_URL) -> bool:
+        request = requests.get(API_STATUS)
+        if request.status_code == HTTP_OK:
+            return True
+        return False
+
+    def authenticate(username, password, url=DEFAULT_URL) -> bool:
+        auth_headers = {"x-api-user": username, "x-api-key": password}
+        request = requests.get(API_USER_PROFILE, headers=auth_headers)
+        if request.status_code == HTTP_OK:
+            return True
+        return False
+
+    is_valid = await hass.async_add_executor_job(check_api_status, data[CONF_URL])
+    if is_valid is False:
+        raise CannotConnect()
+
+    is_valid = await hass.async_add_executor_job(
+        authenticate, data[CONF_API_USER], data[CONF_API_KEY], data[CONF_URL]
+    )
+    if is_valid is False:
+        raise InvalidAuth()
+    return {"title": f"{data.get('name', 'Default username')}"}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for habitica."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+
+        if not user_input:
```

Comment:
Invert this check and do the validation if true. End the step with a single show form call with possible errors.